src/opensynth/models/faraday/gaussian_mixture/metrics.py

```diff
@@ -167,17 +169,25 @@ def update(
                 covars = (
                     responsibilities[:, i].unsqueeze(1) * component_diff
                 ).T.matmul(component_diff)
-                self.covariance_sum[i].add_(covars)
+
+                # Add regularization to the diagonal of the covariance matrix
+                regularization = self.reg * torch.eye(
+                    self.num_features, device=covars.device
+                )
+                covars_regularized = covars + regularization
+
+                self.covariance_sum[i].add_(covars_regularized)
 
     def compute(self) -> torch.Tensor:
         if self.covariance_type == "diag":
             return (
-                self.covariance_sum / self.component_weights.unsqueeze(-1)
+                self.covariance_sum
+                / self.component_weights.unsqueeze(-1).add_(1e-16)
                 + self.reg
             )
         if self.covariance_type == "spherical":
             return (
-                self.covariance_sum / self.component_weights
+                self.covariance_sum / self.component_weights.add_(1e-16)
                 + self.reg * self.num_features
             )
         if self.covariance_type == "tied":
```

Comment:
If we are not using covariance_types other than "full", maybe we should deleted references to them given we don't know if using them would be successful...

Reply:
Yep we could, do you know if we did any experiments using non "full" covariances when using sklearn? maybe @shengy90 would know if this is something we would want to try out in future.

Reply:
Full is more commonly used, allows each cluster to have it's own covariance matrix. Diagonal is also sometimes used when we require all clusters to share the same diagonals in the covariance matrix, but IDK why we'd want that. If full covariance matrix leads to 'overfitting' then normal procedure is to reduce number of clusters, rather than forcing all clusters to share the same covariance matrix (tied), or forcing all clusters to be spherical.

If we're worried about the correctness of the outputs then I think we can just implement `full` and comment in the code that spherical, tied and diag aren't supported yet (and let public raise a PR if they so wish).